lib/modelExtends.js

```diff
@@ -44,14 +48,20 @@ module.exports = function(Sequelize) {
 			}, sequelize.options.hierarchy || {}, options);
 
 			_.defaults(options, {
-				foreignKey: options.as + Utils.uppercaseFirst(options.primaryKey),
-				throughKey: this.name + Utils.uppercaseFirst(options.primaryKey),
-				throughForeignKey: Utils.singularize(options.ancestorsAs) + Utils.uppercaseFirst(options.primaryKey),
+				foreignKey: patches.underscoredIf(options.as + Utils.uppercaseFirst(options.primaryKey), underscored),
+				throughKey: patches.underscoredIf(this.name + Utils.uppercaseFirst(options.primaryKey), underscored),
+				throughForeignKey: patches.underscoredIf(Utils.singularize(options.ancestorsAs) + Utils.uppercaseFirst(options.primaryKey), underscored),
 				through: this.name + Utils.singularize(options.camelThrough ? Utils.uppercaseFirst(options.ancestorsAs) : options.ancestorsAs)
 			});
 
+			var genThroughTableName;
+			if (underscoredAll)
```

Comment:
No haha, that's not what I meant. There's already a `throughTable` option which (even with this PR) behaves as you described.

You proposed to use `this.sequelize.options.define.underscoredAll` instead of `this.options.underscoredAll` for applying underscoring to the `throughTable`. To that, I replied:

> That sounds reasonable, I think it'd need to be documented in that case though.
> 
> If we were to do this, this.sequelize.options.define is empty by default. Should I let the throughTable name be underscored if this.sequelize.options.define has property underscoredAll and the properties value is true?